src/hotspot/share/opto/parse1.cpp

```diff
@@ -507,34 +494,44 @@ Parse::Parse(JVMState* caller, ciMethod* parse_method, float expected_uses)
 
   // Do some special top-level things.
   if (depth() == 1 && C->is_osr_compilation()) {
+    _tf = C->tf();     // the OSR entry type is different
     _entry_bci = C->entry_bci();
     _flow = method()->get_osr_flow_analysis(osr_bci());
-    if (_flow->failing()) {
-      assert(false, "type flow analysis failed for OSR compilation");
-      C->record_method_not_compilable(_flow->failure_reason());
+  } else {
+    _tf = TypeFunc::make(method());
+    _entry_bci = InvocationEntryBci;
+    _flow = method()->get_flow_analysis();
+  }
+
+  if (_flow->failing()) {
+    assert(false, "type flow failed during parsing");
+    C->record_method_not_compilable(_flow->failure_reason());
 #ifndef PRODUCT
       if (PrintOpto && (Verbose || WizardMode)) {
-        tty->print_cr("OSR @%d type flow bailout: %s", _entry_bci, _flow->failure_reason());
+        if (_entry_bci != InvocationEntryBci) {
```

Comment:
```suggestion
        if (C->is_osr_compilation()) {
```

Reply:
This is not exactly the same. we want to execute this statement if and only if (C->is_osr_compilation() && depth() == 1)!  

```
tty->print("OSR @%d ", _entry_bci);
```

for OSR, Parsing is still recursive. In callee's Parse,  _entry_bci is InvocationEntryBci but C->is_osr_compilation() is true. 

if _entry_bci != InvocationEntryBci is true, it's only possible from 'depth() == 1 && C->is_osr_compilation())' above.

Reply:
Got it, thanks for the clarification!